### PR TITLE
Stable GM memory usage during constant redelivery

### PR DESCRIPTION
**Pull request by gerhard**
In high throughput scenarios, e.g. `basic.reject` or `basic.nack`,
messages which belong to a mirrored queue and are replicated within a GM
group, are quickly promoted to the old heap. This means that garbage
collection happens only when the Erlang VM is under memory pressure,
which might be too late. When a process is under pressure, garbage
collection slows it down even further, to the point of RabbitMQ nodes
running out of memory and crashing. To avoid this scenario, We want the
GM process to garbage collect binaries regularly, i.e. every 250ms. The
variable queue does the same for a similar reason:
rabbitmq/rabbitmq-server#289

Initially, we wanted to use the number of messages as the trigger for
garbage collection, but we soon discovered that different workloads
(e.g. small vs large messages) would result in unpredictable and
sub-optimal GC schedules.

Before setting `fullsweep_after` to `0`, memory usage was 2x higher (400MB
vs 200MB) and throughput was 0.1x lower (18k vs 20k). With this
`spawn_opt` setting, the general collection algorithm is disabled,
meaning that all live data is copied at every garbage collection:
http://erlang.org/doc/man/erlang.html#spawn_opt-3

The RabbitMQ deployment used for testing this change:

* AWS, c4.2xlarge, bosh-aws-xen-hvm-ubuntu-trusty-go_agent 3421.11
* 3 RabbitMQ nodes running OTP 20.0.1
* 3 durable & auto-delete queues with 3 replicas each
* each queue master was defined on a different RabbitMQ node
* every RabbitMQ node was running 1 queue master & 2 queue slaves
* 1 consumer per queue with QOS 100
* 100 durable messages @ 1KiB each
* `basic.reject` operations

```
| Node   | Message throughput   | Memory usage   |
| ------ | -------------------- | -------------- |
| rmq0   | 12K - 20K msg/s      | 400 - 900 MB   |
| rmq1   | 12K - 20K msg/s      | 500 - 1000 MB  |
| rmq2   | 12K - 20K msg/s      | 500 - 800 MB   |
```

### CODE REVIEW

**Comment by gerhard** [comment]
Letting the benchmark run for 12h before it's ready to merge.

**Comment by lukebakken** [comment]
See #290 and #339 for `garbage_collect()` in `rabbit_variable_queue`

File: src/gm.erl [hunk 1]
```diff
@@ -883,14 +888,24 @@ handle_msg({activity, _NotLeft, _Activity}, State) ->
...
+
+ensure_timers(State) ->
+    ensure_force_gc_timer(ensure_broadcast_timer(State)).
 
-flush_timeout(#state{broadcast_buffer = []}) -> hibernate;
+flush_timeout(#state{broadcast_buffer = []}) -> infinity;
 flush_timeout(_)                             -> 0.
 
+ensure_force_gc_timer(State = #state { force_gc_timer = TRef })
+  when TRef =/= undefined ->
```
Comment (lukebakken) [inline src/gm.erl]: `TRef =/= undefined` could be `is_reference(TRef)` but it doesn't really matter 😄

Comment (dumbbell) [inline src/gm.erl]: You're right, we'll change that. Thanks!

**Comment by gerhard** [comment]
Having run the benchmark for 18h, we are confident that high message redelivery rates for mirrored queues no longer affect the cluster stability:

image

Even though there is a noticeable memory fluctuation after 11h, it's not significant. Memory usage grows from 400MB to 1000MB, it remains stable for 4h, then drops back to 400MB before another short increase. Eventually, memory returns back to 400MB and remains stable for the rest of the 18h. We observe the same behaviour on all nodes. Since the memory returns back to normal and since the cluster is stable throughout, we are happy to just take note of this and not investigate further.

**Comment by gerhard** [comment]
@michaelklishin @dcorbacho ready to review & merge

**Comment by gerhard** [comment]
We've left **rmq-148892851** around, in case you need to use it for further benchmarks.

**Comment by langyxxl** [comment]
When in openstack large deployment, 1000 compute node ovs-agent  will create 10K ha-queue.
This gc every 250ms will cause cpu usage very high. Even these queue is empty.

https://groups.google.com/forum/?nomobile=true#!topic/rabbitmq-users/6jGtaHINmNM

**Comment by michaelklishin** [comment]
@langyxxl please keep discussions to the mailing list. Thank you.

**Comment by gerhard** [comment]
TL; DR the Erlang VM spends 44.80% in `busy_wait`. Setting `+sbwt none` via `RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS` significantly reduces CPU utilisation - a few % in both user & system - when there are 10k ha-all queues.

[Read more and discuss on the mailing list thread](https://groups.google.com/d/msg/rabbitmq-users/6jGtaHINmNM/rc1rR1PqAwAJ)

cc @michaelklishin @dcorbacho
